File: dfs/new/remove_boxes.py
class Solution:
    def removeBoxes(self, boxes):
        def dfs(l, r, k):
            if l > r:
                return 0
            if (l, r, k) not in memo:
                while r > l and boxes[r] == boxes[r-1]:
                    k += 1
                    r -= 1
                while l < r and boxes[l] == boxes[r]:
                    k += 1
                    l += 1
                res = (k+1)**2 + dfs(l, r-1, 0)
                for i in range(l+1, r-1):
                    if boxes[i] == boxes[r]:
                        res = max(res, dfs(l, i, k+1) + dfs(i+1, r-1, 0))
                memo[l, r, k] = res
            return memo[l, r, k]
        memo = {}
        return dfs(0, len(boxes) - 1, 0)

# I found this solution in the score distribution.
# 89% 108ms
# memo[l, r, k] is the maximum points possible by removing the boxes[ l: r ] with k boxes outside boxes[ l : r ] that shares the same color as boxes[r].Because our recursion starts with dfs(0,len(boxes)-1, 0), so k doesn't include boxes[r] but the result does: res = (k+1)**2 + dfs(l, r-1, 0)

# The core recursion formula

#   if boxes[i] == boxes[r]:
#             res = max(res, dfs(l, i, k+1) + dfs(i+1, r-1, 0))
# We want all boxes[i] that has the same color as boxes[r] to be removed together, dfs(l, i, k+1) does just that.

# [1,3,1,3,3,1]
# In this example, i = 2, boxes[2] = boxes[5] = 1, dfs(1, 2, 2) => res = (k+1) ** 2 + dfs(1, 1,0) = 9 + 1 = 10

# Example: [1, 3, 1, 2, 1, 3, 1]
# memo:
# {(3, 3, 0): 1, (3, 4, 1): 5, (2, 5, 1): 9, (1, 1, 0): 1, (1, 2, 2): 10, (3, 4, 0): 2, (3, 5, 0): 3, (1, 2, 0): 2, (1, 3, 0): 3, (1, 2, 3): 17, (1, 4, 2): 18, (5, 5, 0): 1, (1, 6, 1): 19}
# Final ans: 19
# Let's focus on how we got (2, 5, 1): 9. 2th to 5th subarray is [1,2,1,3].
# boxes[r] = 3 and there is exactly one 3 in the rest of the array, that's why we have k = 1.

#     1. l=0, r=len(boxes)-1=6, because they are equal, l +=1, r -=1
#     2. l = 1, r = 5, boxes[l]= boxes[r] = 3
#     3. l = 2, r = 4, boxes[l] = boxes[r] = 1
#     4. l = r = 3, boxes[l] = 2
#     5. We return to the caller l = 2, r = 4, res = 2 ** 2 + 1 = 5, since l + 1= 3 = r - 1, we return to l = 1, r= 5, res = 2 ** 2 + 5 = 9
#         for i in range(2, 4):
#         if boxes[i] == boxes[5] = 3:
#         None of the box meets the condition, thus memo[2,5,1] = res = 9.
# Why is l = 2 instead of l = 1 as we have passed in, because we increment l when boxes[l] = boxes[r], remember that k is the number of boxes outside of boxes[ l : r] with the same color as boxes[r].

File: dfs/new/remove_boxes.cpp
class Solution {
public:
    int mem[100][100][100]; // initialized to 0, mem[left][right][k] means value from boxes[left]~boxes[right] followed by 
    // k same color boxes. Follow does not mean strictly consecutive boxes, for example, [1, 3, 2, 3, 4], 3 can be 
    // followed by the other 3 because we can remove 2 first
    
    int removeBoxes(vector<int>& boxes) {
        return DFS(boxes,0,boxes.size()-1,0);
    }
    
    int DFS(vector<int>& boxes, int l,int r,int k){
        if (l>r) return 0; 
        if (mem[l][r][k]) return mem[l][r][k]; // if we have calculated this DFS result, return it
        
        mem[l][r][k] = DFS(boxes,l,r-1,0) + (k+1)*(k+1); // box[l][r] result is box[l][r-1]+(k+1)^2
        for (int i=l; i<r; i++) // go through each box from left
            if (boxes[i]==boxes[r]) // check for same color box as boxes[r]
                mem[l][r][k] = max(mem[l][r][k], DFS(boxes,l,i,k+1) + DFS(boxes,i+1,r-1,0)); // if we found same color box,
                // then we have a chance to get a higher value by group boxes[l]~boxes[i] and boxes[r] together, plus the 
                // value from boxes[i+1]~boxes[r-1]
        return mem[l][r][k];
    }
};